src/cpp/statistics/rtps/monitor-service/MonitorServiceListener.cpp

```diff
@@ -86,6 +87,109 @@ void MonitorServiceListener::on_writer_change_received_by_all(
     static_cast<void>(change);
 }
 
+bool MonitorServiceListener::on_incompatible_qos_matching(
+        const fastdds::rtps::GUID_t& local_guid,
+        const fastdds::rtps::GUID_t& remote_guid,
+        const fastdds::dds::PolicyMask& incompatible_qos_policies) const
+{
+    bool ret = true;
+
+    // Convert the PolicyMask to a vector of policy ids
+    std::vector<uint32_t> incompatible_policies;
+    for (uint32_t id = 1; id < dds::NEXT_QOS_POLICY_ID; ++id)
+    {
+        if (incompatible_qos_policies.test(id))
+        {
+            incompatible_policies.push_back(id);
+        }
+    }
+
+    if (!incompatible_policies.empty())
+    {
+        std::lock_guard<std::mutex> lock(monitor_srv_->extended_incompatible_qos_mtx_);
+
+        // Check if the local_guid is already in the collection. If not, create a new entry
+        auto local_entity_incompatibilites =
+                monitor_srv_->extended_incompatible_qos_collection_.insert({local_guid, {}});
+
+        bool first_incompatibility_with_remote = false;
+
+        // Local entity already in the collection (has any incompatible QoS with any remote entity)
+        if (!local_entity_incompatibilites.second)
+        {
+            // Check if the local entitiy already had an incompatibility with this remote entity
+            auto it = std::find_if(
+                local_entity_incompatibilites.first->second.begin(),
+                local_entity_incompatibilites.first->second.end(),
+                [&remote_guid](const ExtendedIncompatibleQoSStatus_s& status)
+                {
+                    return to_fastdds_type(status.remote_guid()) == remote_guid;
+                });
+
+            if (it == local_entity_incompatibilites.first->second.end())
+            {
+                // First incompatibility with that remote entity
+                first_incompatibility_with_remote = true;
+            }
+            else
+            {
+                // Already had an incompatibility with that remote entity
+                // Check if the any of the policies were already stored
+                for (auto& policy : incompatible_policies)
+                {
+                    if (std::find(it->current_incompatible_policies().begin(),
+                            it->current_incompatible_policies().end(),
+                            policy) == it->current_incompatible_policies().end())
+                    {
+                        it->current_incompatible_policies().emplace_back(policy);
+                        monitor_srv_->push_entity_update(local_guid.entityId, StatusKind::EXTENDED_INCOMPATIBLE_QOS);
+                    }
+                }
```

Comment:
Could we just assign `incompatible_policies` to `it->current_incompatible_policies()`?

Reply:
Since we always compute all qos incompatibities in EDP and we are only interested in the current ones, we can do it. Nice optimization !